File: server/Application/Services/CardRanking.cs
using Antiphon.Server.Domain.Entities;
using Antiphon.Server.Domain.Enums;

namespace Antiphon.Server.Application.Services;

/// <summary>
/// Single owner of effective urgency, quadrant and rank. Every sort site calls
/// <see cref="Rank(CardImportance, CardUrgency, DateTime?, DateTime, int)"/>; there is no persisted
/// Rank column because it would go stale with the clock.
/// </summary>
/// <remarks>
/// <para><c>rank = 13 − (3·importance + 2·effectiveUrgency)</c>, lower sorts first. The weights
/// prefer a card that changes how everything else gets done over one more feature
/// (<c>Critical/Normal</c> beats <c>Normal/Now</c>), while a standing-red <c>Low/Now</c> still
/// beats the undifferentiated default <c>Normal/Normal</c>.</para>
///
/// <para>The <c>blockedDependants</c> argument on Rank is the CARD-0100 extension point: accepted
/// and ignored until relationships land and that card decides whether a blocker feeds rank or
/// stays a separate column in the reading order.</para>
/// </remarks>
public static class CardRanking
{
    public static readonly TimeSpan NowDueWindow = TimeSpan.FromDays(3);
    public static readonly TimeSpan SoonDueWindow = TimeSpan.FromDays(14);

    public static CardUrgency EffectiveUrgency(CardUrgency stored, DateTime? dueAt, DateTime now)
    {
        var implied = ImpliedByDueAt(dueAt, now);
        return stored >= implied ? stored : implied;
    }

    public static CardUrgency EffectiveUrgency(Card card, DateTime now) =>
        EffectiveUrgency(card.Urgency, card.DueAt, now);

    public static CardQuadrant Quadrant(CardImportance importance, CardUrgency effectiveUrgency)
    {
        var important = importance is CardImportance.High or CardImportance.Critical;
        var urgent = effectiveUrgency != CardUrgency.Normal;
        return (important, urgent) switch
        {
            (true, true) => CardQuadrant.DoFirst,
            (true, false) => CardQuadrant.Schedule,
            (false, true) => CardQuadrant.Clear,
            (false, false) => CardQuadrant.Someday
        };
    }

    public static CardQuadrant Quadrant(Card card, DateTime now) =>
        Quadrant(card.Importance, EffectiveUrgency(card, now));

    /// <summary>
    /// Lower sorts first. <paramref name="blockedDependants"/> is accepted and ignored until
    /// CARD-0100 wires relationships.
    /// </summary>
    public static int Rank(
        CardImportance importance,
        CardUrgency urgency,
        DateTime? dueAt,
        DateTime now,
        int blockedDependants = 0)
    {
        _ = blockedDependants;
        var effective = EffectiveUrgency(urgency, dueAt, now);
        return 13 - (3 * (int)importance + 2 * (int)effective);
    }

    public static int Rank(Card card, DateTime now, int blockedDependants = 0) =>
        Rank(card.Importance, card.Urgency, card.DueAt, now, blockedDependants);

    /// <summary>
    /// Maps the 0–5 tracker adapter scale onto <see cref="CardImportance"/>:
    /// 5 → Critical, 4 → High, 2–3 → Normal, 1 → Low, 0 → Normal.
    /// </summary>
    public static CardImportance FromTrackerScale(int trackerScale) => trackerScale switch
    {
        >= 5 => CardImportance.Critical,
        4 => CardImportance.High,
        3 or 2 => CardImportance.Normal,
        1 => CardImportance.Low,
        _ => CardImportance.Normal
    };

    /// <summary>Earliest due date first; null sorts last.</summary>
    public static DateTime DueAtSortKey(DateTime? dueAt) => dueAt ?? DateTime.MaxValue;

    private static CardUrgency ImpliedByDueAt(DateTime? dueAt, DateTime now)
    {
        if (dueAt is null)
            return CardUrgency.Normal;

        var remaining = dueAt.Value - now;
        if (remaining <= NowDueWindow)
            return CardUrgency.Now;
        if (remaining <= SoonDueWindow)
            return CardUrgency.Soon;
        return CardUrgency.Normal;
    }
}

File: client/src/features/board/cardRanking.ts
export type CardImportance = 'Low' | 'Normal' | 'High' | 'Critical'
export type CardUrgency = 'Normal' | 'Soon' | 'Now'
export type CardQuadrant = 'DoFirst' | 'Schedule' | 'Clear' | 'Someday'

export const IMPORTANCE_VALUES: readonly CardImportance[] = ['Low', 'Normal', 'High', 'Critical']
export const URGENCY_VALUES: readonly CardUrgency[] = ['Normal', 'Soon', 'Now']
export const QUADRANT_ORDER: readonly CardQuadrant[] = ['DoFirst', 'Schedule', 'Clear', 'Someday']

const IMPORTANCE_WEIGHT: Record<CardImportance, number> = {
  Low: 0,
  Normal: 1,
  High: 2,
  Critical: 3,
}

const URGENCY_WEIGHT: Record<CardUrgency, number> = {
  Normal: 0,
  Soon: 1,
  Now: 2,
}

export const NOW_DUE_MS = 3 * 86_400_000
export const SOON_DUE_MS = 14 * 86_400_000

export function effectiveUrgency(
  stored: CardUrgency,
  dueAt: string | null | undefined,
  now: Date,
): CardUrgency {
  const implied = impliedByDueAt(dueAt, now)
  return URGENCY_WEIGHT[stored] >= URGENCY_WEIGHT[implied] ? stored : implied
}

export function quadrant(importance: CardImportance, effective: CardUrgency): CardQuadrant {
  const important = importance === 'High' || importance === 'Critical'
  const urgent = effective !== 'Normal'
  if (important && urgent) return 'DoFirst'
  if (important) return 'Schedule'
  if (urgent) return 'Clear'
  return 'Someday'
}

export function rank(
  importance: CardImportance,
  urgency: CardUrgency,
  dueAt: string | null | undefined,
  now: Date,
): number {
  const effective = effectiveUrgency(urgency, dueAt, now)
  return 13 - (3 * IMPORTANCE_WEIGHT[importance] + 2 * URGENCY_WEIGHT[effective])
}

function impliedByDueAt(dueAt: string | null | undefined, now: Date): CardUrgency {
  if (!dueAt) return 'Normal'
  const remaining = new Date(dueAt).getTime() - now.getTime()
  if (Number.isNaN(remaining)) return 'Normal'
  if (remaining <= NOW_DUE_MS) return 'Now'
  if (remaining <= SOON_DUE_MS) return 'Soon'
  return 'Normal'
}
